### PR TITLE
Turn Order Amendments

### LINKED ISSUE #63: Items Modifying Speed
Implement these speed-affecting items to decide if the raider or the raid boss moves first:

- Choice Scarf (1.5x)
- Iron Ball (.5x)
- Power Items (.5x)
- Lagging Tail (last in priority bracket)
- Full incense (last in priority bracket, not currently available)

### PR DESCRIPTION
**Pull request by Ericzklm**
Choice Scarf, Iron Ball, Power Items, Lagging Tail affect turn order

Closes #63

### CODE REVIEW

**Comment by Ericzklm** [comment]
Added a stage after pulling a Pokemon's speed: check the item and modify the speed after pulling the modified speed stat and before getting the field effects.

Let me know if you want this implemented some other way.

**Comment by Ericzklm** [comment]
Implemented speed modifiers for:

- paralysis
- items
    - choice scarf
    - iron ball
    - power items
    - lagging tail
    - full incense
- abilities
    - unburden *
    - slow start
    - quick feet
- field related effects
    - chlorophyll
    - sand rush
    - slush rush
    - swift swim
    - surge surfer

I kept the implementations consistent with damage-calc meaning some abilities check for abilityOn, others don't. This is why unburden does not currently work because it uses abilityOn for the speed modifier and abilityOn for unburden isn't set until results are calculated.

**Comment by theastrogoth** [comment]
The issue with Unburden is that we don't have items that are triggered before a turn (e.g. Grassy Seed), right? If you wanted to activate Unburden with a Focus Sash, wouldn't `abilityOn` be set appropriately for the following turn?

Oh, only the user is checked for setting Unburden in `RaidMove` 😅. The target(s) should be checked as well to make sure it gets activated.

With abilities like Seed Sower, I get the feeling we're going to have to repeat checks for item use and ability activation in a few more places.

**Comment by Ericzklm** [comment]
Implemented Gale Wings and the priority portion of Prankster (no move fail checks)
I'm not sure how to handle the list of moves that are affected by Prankster, so for right now I enumerated all of them and put them in a JSON.

Moved Protosynthesis and Quark Drive stat boosts to check for all Pokemon for each RaidTurn instead of RaidMove results. Let me know if this breaks anything.

http://localhost:3000/tera-raid-builder#H4sIAAAAAAAAA7WT32/aMBDH/5XIe9kkTyL82CrexqAak9i6tm8oDya5BA/Hjs4Oalbxv+/shFKgexookXU+++77yd3lmeVszNLf1mjGmWPj5bLHmRYlsIR7U2adkaJ0dMVCanQmsJnlOaTOkguNUv5SzFltAedTn0lgAS6YpnLSaOtv9Dkr0NQVeUuzhbnODZkrY+1iv23zaOPAp04RMhlEKrOBsoWsUXtPyGQ7urXPKdyG1gxyz1mJsGZhhU6dUIF130f3V1JJ15AlHZTBT8n9CWy9ggyrgi0orwsoHpsKOni7J6+Vk5WSgD7uyaFYhNMk4WzLxs+MavqJM9a9y+C44cR8L2QWTSgFHdxDIUHBRpL9/oeJvrRoHxjXtVKcfROYEWgI/kzBL0+y2zsH8clLR3Gv1yZYJsFeJvv7NzxuCQCjdzGpztHoaFLrLFToVy1wE01Rbv3u69pUCqIJIPpq/TcMu0WAP/BxSunOwfoHsD6p3Umd1piupW/+TNHQoUyjh5oG7GLVOeR9pG8UJHWONThgDYjkJwpd1FiT+dCUK2ms9I2cGGNpUqKZBiwuU6wJcW2iCYLYvIE1PGANSe67LArVVHWen00SuwMnGnHBNi6kdc1LyTi7VVIXrxiTbtxHvAsJu2EL5n8ToorpLLgHPBfKQrey0jdht0c5julR1Jsx4inE7KNGNEtHuq9LSQ39R5aLKA+OldvSULeuqTk81jxpz/CqlR6dfO/hD79Si5MwYru/MYe3G8IGAAA=

Quark Drive Testing:
- Turn 1: Iron Bundle outspeeds Regieleki from QP boost to speed thanks to Quark Drive + Electric Surge
- Turn 2: Oranguru uses Symbiosis to pass Booster Energy to Iron Bundle. Booster Energy is not consumed
- Turn 3: Jigglypuff flings Petaya Berry making SpAtk the highest stat for QP
- Turn 4: Jigglypuff changes the terrain, causing Iron Bundle to consume the Booster Energy, now boosting SpAtk
- Turn 5: Iron Bundle underspeeds Regieleki because the QP boost is no longer on Speed

Edit: I think there is an issue after turn 3: I believe the QP boost should only be changed if it's reactivated but adding another move where Iron Bundle attacks between turns 3 and 4 shows that the SpAtk boost immediately goes into effect.

**Comment by theastrogoth** [comment]
It looks like we might need to store the method of QP activation. If a booster energy was consumed, then we can skip all checks and keep the current boost. If not, we can handle separately the cases where Sun/Electric Terrain begins and ends

File: src/raidcalc/RaidTurn.ts [hunk 1]
```diff
@@ -139,9 +144,35 @@ export class RaidTurn {
...
+        this.raidState.raiders.map((raider, index) => {
+            if (raider.ability === "Protosynthesis" || raider.ability === "Quark Drive") {
+                if (
+                    (this.raidState.fields[index].weather?.includes("Sun") && raider.ability === "Protosynthesis") ||
+                    (this.raidState.fields[index].terrain?.includes("Electric") && raider.ability === "Quark Drive")
+                ) { // Sun / Electric Terrain has priority over Booster Energy, Booster Energy is not consumed if QP is already active
+                    raider.abilityOn = true;
+                    raider.boostedStat = undefined;
+                    const qpStat = getQPBoostedStat(raider) as StatIDExceptHP;
+                    raider.boostedStat = qpStat;
```
Comment (theastrogoth) [inline src/raidcalc/RaidTurn.ts:166]: This seems to be the culprit for boost changing while sun/e-terrain is active.

**Comment by theastrogoth** [comment]
I've worked a bit on getting QP to work as intended, as well as on trying to make sure T0 ability/item triggers actually happen. 

QP just needed a few tweaks in RaidTurn. Your test example seems to work

T0 stuff gets handled by getting results from "dummy moves" (where everyone uses "(No Move)") at the start of a battle in RaidBattle. I've tested with a Grassy Seed and Unburden, and things seem to be okay:

http://localhost:3000/tera-raid-builder#H4sIAAAAAAAAA82TTY/TMBCG/0pkLiD50E9Y9dZSYBdp99AtcIhymCST1tTxRLZTWq32vzN20m1BPYC4oEajyXg+nnnrPIlKzETx3ZERUngxS9OBFAZqFJkMrip7p7DKc4rDgkwJ9vihqrDwjkOWtA5JQylah/ZuGTqB3aCPLjVekXEhYyTFxlLbcLSmPd6ZitjNybn702vXx5DH0LqwWKo4pKEd1h1ka02IxE6up9uGnuB3bEusAmcD0ZbRYj+dUVH0+3F+rrTyR/aUxzrGuXk4wX2YoKLVuEcd5qKF9bHBHt6dyFvtVaMV2lB38Bbu42mWSbEXsyfBmr6VQvRPGgM3kplXoMpkwS34YKW0hpwoUHyy4NwxeWxZQiFNq7UUn1ljBo3F77j45Zc9n4Lj4W8PHw0HPOeBbA28QSrWsMNkST+CjN+IyuQW6jqCzwtLOXhVBJilbetkgfxqNizceMpdUm6Ta0xWBCF/Da0J12HByHlrXeeXTK2h5t17qhs57PZEm7wacs6jQXA6jvxiuLBEc7ExYvi7Fhb2+Fe7DvpVo1bp5TJZjKcXQKMz0CjsCsXW5W2gmFuqIfmKSp9UX5AuX0BG0xFDdPaPhO9pblE3LCRrbcorPOMzz5gh3pPdq51Rm22Q9PUDJfPulr45Qa1QwwH/nWuJFW14xsdwKdYI9ppakzPd5P+hy/rPaCr7/Pg2+fUa87Xo4mNZgXbYW1ErI7jF89WiAZddLYJDLIqzmeD5Jwi4Zbs5BQAA

I've also added a new section to the Calc Results display to hopefully make it a little easier to see when things are getting activated appropriately.

**Comment by Ericzklm** [comment]
Updated test case:
http://localhost:3000/tera-raid-builder#H4sIAAAAAAAAA8VUUU/bMBD+K5H3skme1NACXd/oAI1JbAx4q/Jgkkvw4trR2anIEP99ZyehTcue1m5KZJ3Pd9/35e7iZ5azGUt/WqMZZ47NFosRZ1osgSXcmzLrjBSloxALqdGZwOYizyF1llxolPJBMWe1Bbw690gCC3DBNJWTRlsfccRZgaauyLs0K7jSuSHzwVh73W9bHG0ceOgUIZOBpDIlLFuRNWrvCUi2U/foMYUrac0g9zorEdYsrNCxk1Rg3fdR/INU0jVkSQfL4CdwfwIrzyDDqmAFyvMCivumgk687ZXXyslKSUCf9+RQXIfTJOFsxWbPjGp6whnr3kVwTDlpvhUyi+YEQQe3UEhQUEqy338z0Vkr7QPjulaKsy8CMxIakk8p+fVJXnrnON566SgejVqARRLsRdLHT3ncKgCM3sXEeoVGR/NaZ6FCP2qBZXSOcuV3nx9NpSCaA6Kv1l+LYZcI8As+nge4s6Jtw67Eo7XEIwq8kTqtMX2UfgwuFI0fyjS6q2nU9lanNe49fa0gql1Z47WsMSn5jkIXNdZk3jXLB2ms9C2dG2NpZqILDVjsp2xz0lVGcwRRviFrspY1IbqvsihUU9V5vjNT7AacaMQeG3otrWteS8bZpZK62NCYdIN/zLuUsJu0wvwPQ6piOgvuMc+FstCtbOmb8NJLGeaMKOvNHPEUcvqsY5qlAe9mKamhf0DZC/N4yNyWhrp1SM6TLc6N/+2wZZ4MibfmYnJQ7uP/9dGnQ+L+Qjsw63TIunNx0cgfkv7Tv652Em6Sl98TYWNjswgAAA==

- Turn 1: Iron Bundle outspeeds Regieleki from QP boost to speed thanks to Quark Drive + Electric Surge
- Turn 2: Oranguru uses Symbiosis to pass Booster Energy to Iron Bundle. Booster Energy is not consumed
- Turn 3: Jigglypuff flings Petaya Berry making SpAtk the highest stat for QP
- Turn 4: Iron Bundle still maintains speed QP boost despite SpAtk now being highest stat
- Turn 5: Jigglypuff changes the terrain, causing Iron Bundle to consume the Booster Energy, now boosting SpAtk
- Turn 6: Iron Bundle underspeeds Regieleki because the QP boost is no longer on Speed
- Turn 7: Iron Bundle uses Agility to make Speed once again the highest stat for the QP boost
- Turn 8: Electric Surge is set up again but QP boost was permanently set already and does not change
- Turn 9: Iron Bundle outspeeds Regieleki but the QP boost remains on SpAtk

**Comment by Ericzklm** [comment]
I believe thats all the speed modifiers covered!

Prankster failures should probably be addressed in a separate PR where move failures in general are addressed.

**Pull request merged**